crates/xtask-bump-check/src/xtask.rs

```diff
@@ -0,0 +1,338 @@
+        .unwrap();
+
+    let mut changed_members = HashMap::new();
+
+    for delta in diff.deltas() {
+        let mut insert_changed = |path: &Path| {
+            if !member_dirs.iter().any(|dir| path.starts_with(dir)) {
+                return;
+            }
+            let Some(member) = path.components().nth(1) else {
+                log::trace!("skipping {path:?}, not a valid member path");
+                return;
+            };
+            let name = member.as_os_str().to_str().unwrap();
+            let Some((name, pkg)) = ws_members.get_key_value(name) else {
+                log::trace!("skipping {name:?}, not found in workspace");
+                return;
+            };
+            if pkg.publish() == &Some(vec![]) {
+                log::trace!("skipping {name}, `publish = false`");
+                return;
+            }
+            changed_members.insert(*name, *pkg);
+        };
+
+        insert_changed(delta.old_file().path().unwrap());
+        insert_changed(delta.new_file().path().unwrap());
+    }
+
+    Ok(changed_members)
+}
+
+/// Compares version against published crates on crates.io.
+///
+/// Assumption: We always release a version larger than all existing versions.
+fn check_crates_io<'a>(
+    config: &Config,
+    changed_members: &HashMap<&str, &'a Package>,
+    needs_bump: &mut Vec<&'a Package>,
+) -> CargoResult<()> {
+    let source_id = SourceId::crates_io(config)?;
+    let mut registry = PackageRegistry::new(config)?;
+    let _lock = config.acquire_package_cache_lock()?;
+    registry.lock_patches();
+    config.shell().status(
+        "BumpCheck",
+        format_args!("compare against `{}`", source_id.display_registry_name()),
+    )?;
+    for member in changed_members.values() {
+        let name = member.name();
+        let current = member.version();
+        let version_req = format!("<={current}");
+        let query = Dependency::parse(name, Some(&version_req), source_id)?;
+        let possibilities = loop {
+            // Exact to avoid returning all for path/git
+            match registry.query_vec(&query, QueryKind::Exact) {
+                task::Poll::Ready(res) => {
+                    break res?;
+                }
+                task::Poll::Pending => registry.block_until_ready()?,
+            }
+        };
+        let max_version = possibilities.iter().map(|s| s.version()).max();
+        if max_version >= Some(current) {
+            needs_bump.push(member);
+        }
+    }
+
+    Ok(())
+}
+
+/// Main entry of `xtask-bump-check`.
+///
+/// Assumption: version number are incremental. We never have point release for old versions.
+fn bump_check(args: &clap::ArgMatches, config: &mut cargo::util::Config) -> CargoResult<()> {
+    let ws = args.workspace(config)?;
+    let repo = git2::Repository::open(ws.root()).unwrap();
+    let base_commit = arg_to_commit(args, &repo, "baseline-rev")?;
+    let head_commit = arg_to_commit(args, &repo, "head-rev")?;
```

Comment:
Can we make this optional and fallback to [HEAD](https://github.com/rust-lang/cargo/pull/12089/files#diff-294d5621c148475db1f52332ac5a5d808b52516e931cbf48883beb6727915ea5R139)?